File: pa1/List.java
//Alejandro Aguilar aaguil10@ucsc.edu


class List{

   private class Node{
      //fileds
      private int data;
      private int index;
      private Node next;
      private Node prev;

      //Constructor
      Node(int _data){
         data = _data;
         next = null;
         prev = null;
         index = -1;
      }
   }


//fileds
private int length;
private Node front;
private Node back;
private Node cursor;

// Constructor
// Creates a new empty list.
List(){
   length = 0;
   front = null;
   back = null;
   cursor = null;

} 

// Access functions
// Returns number of elements in this list.
int length(){
   return length; 
}

// Returns the index of the cursor element in this list, or
// returns -1 if the cursor element is undefined.
int getIndex(){
   if(cursor == null){
      return -1;
   }
   return cursor.index;
}
 
// Returns front element in this list. Pre: length()>0
int front(){
   if(length  <= 0){
      System.err.println("ERROR: In front() length <= 0");
      System.exit(1);
   }
   return front.data;
}

// Returns back element in this List. Pre: length()>0
int back(){
   if(length <= 0){
      System.err.println("ERROR: In back() length <= 0");
      System.exit(1);
   }
   return back.data;
}

// Returns cursor element in this list.
// Pre: length()>0, getIndex()>=0
int getElement(){
   if(length <= 0){
      System.err.println("ERROR: In getElement(), length <= 0");
      System.exit(1);
   }
   if(cursor == null){
      System.err.println("ERROR: In getElement(), cursor == -1");
      System.exit(1);
   }
   return cursor.data;
} 

// Returns true if this List and L are the same integer
// sequence. The cursor is ignored in both lists.
boolean equals(List L){
   boolean flag = true;
   Node N = this.front;
   Node M = L.front;

   if(this.length == L.length){
      while(flag && N != null){
         flag = (N.data == M.data);
         N = N.next;
         M = M.next;
      }
      return flag;
   }else{
      return false;
   }
}



// Manipulation procedures
// Re-sets this List to the empty state.
void clear() {
   length = 0;
   front = null;
   back = null;
   cursor = null;
}

// If 0<=i<=length()-1, moves the cursor to the element
// at index i, otherwise the cursor becomes undefined.
void moveTo(int i){
   if(i < 0 || i > length - 1){
      cursor.index = -1;
   }
   Node curr = front;
   while(curr.index != i){
      curr = curr.next;
   } 
   cursor = curr;
} 

// If 0<getIndex()<=length()-1, moves the cursor one step toward the
// front of the list. If getIndex()==0, cursor becomes undefined.
// If getIndex()==-1, cursor remains undefined. This operation is
// equivalent to moveTo(getIndex()-1).
void movePrev(){
   if(cursor == null){
      return;
   }
   if(cursor.index > length - 1){
      System.err.println("ERROR: In movePrev, cursor index >= length");
      System.exit(1);
   }
   if(cursor.index == 0){
      cursor = null;
      return;
   }
   cursor = cursor.prev;
}

// If 0<=getIndex()<length()-1, moves the cursor one step toward the
// back of the list. If getIndex()==length()-1, cursor becomes
// undefined. If index==-1, cursor remains undefined. This
// operation is equivalent to moveTo(getIndex()+1).
void moveNext() {
   if(cursor == null){
      return;
   }
   if(cursor.index == length){
      cursor = null;
   }
      cursor = cursor.next;
}

// Inserts new element before front element in this List.
void prepend(int data){
   Node tmp = new Node(data);
   if(front == null){
      tmp.index = 0;
      front = tmp;
      back = tmp;
      length++;
      return;
   }
   tmp.next = front;
   front.prev = tmp;
   front = tmp;
   int i = 0;
   Node curr = front;
   while(curr.next != null){
      curr.index = i++;
      curr = curr.next;
   }
   curr.index = i;
   length++;
}

// Inserts new element after back element in this List.
void append(int data){
   Node tmp = new Node(data);
      if(back == null){
      tmp.index = 0;
      front = tmp;
      back = tmp;
      length++;
      return;
   }
   tmp.prev = back;
   tmp.index = back.index + 1;
   back.next = tmp;
   back = tmp;
   length++;
}

// Inserts new element before cursor element in this
// List. Pre: length()>0, getIndex()>=0
void insertBefore(int data){
   if(cursor == null){
      System.err.println("ERROR: In List.insertBefore(), cursor is undefined");
      System.exit(1);
   }
   if(length <= 0){
      System.err.println("ERROR: In List.insertBefore(), length <= 0");
      System.exit(1);
   }

   Node tmp = new Node(data);
   tmp.next = cursor;        //ajusting pointers
   if(cursor != front){
      tmp.prev = cursor.prev;
      cursor.prev.next = tmp;
      cursor.prev = tmp;

   }else{
      cursor.prev = tmp;
      front = tmp;
   }
   
   int i = cursor.index;    //Updating index
   Node curr = tmp;
   while(curr.next != null){
      curr.index = i++;
      curr = curr.next;
   }
   curr.index = i;
   length++;
}

// Inserts new element after cursor element in this
// List. Pre: length()>0, getIndex()>=0
void insertAfter(int data){
   if(cursor == null){
      System.err.println("ERROR: In List.insertAfter(), cursor is undefined");
      System.exit(1);
   }
   if(length <= 0){
      System.err.println("ERROR: In List.insertAfter(), length <= 0");
      System.exit(1);
   }
   Node tmp = new Node(data);
   tmp.prev = cursor;        //ajusting pointers
   if(cursor != back){
      tmp.next = cursor.next;
      cursor.next.prev = tmp;
      cursor.next = tmp;
   }else{
      cursor.next = tmp;
      tmp.index = cursor.index +1;
      back = tmp;
      length++;
      return;
   }
   int i = cursor.index + 1;    //Updating index
   Node curr = cursor.next;
   while(curr.next != null){
      curr.index = i++;
      curr = curr.next;
   }
   curr.index = i;
   length++;
}

// Deletes the front element in this List. Pre: length()>0
void deleteFront(){
   if(length <= 0){
      System.err.println("ERROR: In List.deleteFront(), length <= 0");
      System.exit(1);
   }
   front = front.next;
   front.prev = null;
   int i = 0;   
   Node curr = front;
   while(curr.next != null){
      curr.index = i++;
      curr = curr.next;
   }
   curr.index = i;
   length--;
}

// Deletes the back element in this List. Pre: length()>0
void deleteBack(){
   if(length <= 0){
      System.err.println("ERROR: In List.deleteBack(), length <= 0");
      System.exit(1);
   }
   back = back.prev;
   back.next = null;
   length--;
}

// Deletes cursor element in this List. Cursor is undefined after this
// operation. Pre: length()>0, getIndex()>=0
void delete(){
   if(cursor == null){
      System.err.println("ERROR: In List.delete(), cursor is unidentifed");
      System.exit(1);
   }
   if(length <= 0){
      System.err.println("ERROR: In List.delete(), length <= 0");
      System.exit(1);
   }
   Node tmp = cursor.prev;
   tmp.next = cursor.next;
   int i = tmp.index;
   Node curr = tmp;
   while(curr.next != null){
      curr.index = i++;
      curr = curr.next;
   }
   curr.index = i;
   cursor = null;
   length--;
}



// Other methods
//Overides Object's toString method. Creates a string
//consisting of a space separated sequence of integers
//with front on the left and back on the right. The
//cursor is ignored.
public String toString(){
   String str = "";
   for(Node N = front; N != null; N= N.next){
      str += N.data + " ";
   }
   return str;
}


 // Returns a new list representing the same integer sequence as this
 // list. The cursor in the new list is undefined, regardless of the
 // state of the cursor in this List. This List is unchanged.
 // Returns a new List which is the concatenation of this list
 // followed by L. The cursor in the new list is undefined,
 // regardless of the states of the cursors in this list and L.
 // The states of this list and L are unchanged.

List copy(){
   List myList = new List();
   Node N = this.front;
   
   while(N != null){
      myList.append(N.data);
      N = N.next;
   }
   return myList;

}



}

File: pa2/List.c
//Alejandro Aguilar aaguil10@ucsc.edu

#include<stdio.h>
#include<stdlib.h>
#include "List.h"

typedef struct NodeObj{
   int data;
   struct NodeObj* prev;
   struct NodeObj* next;

} NodeObj;

typedef struct CursorObj{
   Node pointer;
   int index;
} CursorObj;

typedef struct ListObj{
   Node front;
   Node back;
   int length;
   Cursor cruz;
} ListObj;





// Constructors-Destructors ---------------------------------------------------
Node newNode(int dat){
   Node myNode = malloc(sizeof(Node));
   myNode->data = dat;
   myNode->prev = NULL;
   myNode->next = NULL;
   return myNode;
}

Cursor newCursor(void){
   Cursor myCursor = malloc(sizeof(Cursor));
   myCursor->pointer = NULL;
   myCursor->index = -1;
   return myCursor;
}


List newList(void){
   List myList = malloc(sizeof(List));
   myList->front = NULL;
   myList->back = NULL;
   myList->length = 0;
   myList->cruz = newCursor();
   return myList;
}

void freeNode(Node* pN){
   if(pN != NULL && *pN != NULL){
      free(*pN);
      *pN = NULL;
   }
}

void freeCursor(Cursor* pC){
   if(pC != NULL && *pC != NULL){
      free(*pC);
      *pC = NULL;
   }

}

void freeList(List* pL){
   if(pL != NULL && *pL != NULL){
      freeCursor(&(*pL)->cruz); 
      free(*pL);
      *pL = NULL;
    }
}

// Access functions -----------------------------------------------------------
int length(List L){
   if(L == NULL){
      printf("List ERROR: calling length() on a NULL List reference");
      exit(1);
   }
   return L->length;
}

int getIndex(List L){
   if(L == NULL){
      printf("List ERROR: calling getIndex() on a NULL List reference");
      exit(1);
   }
   return L->cruz->index;
}

int front(List L){
   if(L == NULL){
      printf("List ERROR: calling front() on a NULL List reference");
      exit(1);
   }
   if(L->length <= 0){
      printf("List ERROR: calling front() when length <= 0");
      exit(1);
   }
   return L->front->data;
}

int back(List L){
   if(L == NULL){
      printf("List ERROR: calling back() on a NULL List reference");
      exit(1);
   }
   if(L->length <= 0){
      printf("List ERROR: calling back() when length <= 0");
      exit(1);
   }

   return L->back->data;;
}

int getElement(List L){
   if(L == NULL){
      printf("List ERROR: calling getElement() on a NULL List reference");
      exit(1);
   }
   if(L->length <= 0){
      printf("List ERROR: calling getElement() when length <= 0");
      exit(1);
   }
   if(L->cruz == NULL){
      printf("List ERROR: calling getElement() when cruz == NULL");
      exit(1);
   }
   return L->cruz->pointer->data;
}

int equals(List A, List B){
   if(A == NULL){
      printf("List ERROR: calling equals() on a NULL List A reference");
      exit(1);
   }
   if(B == NULL){
      printf("List ERROR: calling equals() on a NULL List B reference");
      exit(1);
   }
   int flag = 1;//bool true
   Node N = A->front;
   Node M = B->front;

   if(A->length == B->length){
      while(flag == 1 && N != NULL){
         flag = (N->data == M->data);
         N = N->next;
         M = M->next;
      }
      return flag;
   }else{
      return 0;
   }
}

// Manipulation procedures ----------------------------------------------------
void clear(List L){
   if(L == NULL){
      printf("List ERROR: calling clear() on a NULL List reference");
      exit(1);
   }
   L->cruz->pointer = NULL;
   L->cruz->index = -1;
      if(L->front != NULL){
         while(L->front->next != NULL){
            deleteFront(L);
         }
         freeNode(&L->front);
         L->front = NULL;
         L->back = NULL;
         L->length = 0;
   }
}

void moveTo(List L, int i){
   if(L == NULL){
      printf("List ERROR: calling moveTo() on a NULL List reference");
      exit(1);
   }
   if(L->cruz->index == -1){
      return;
   }
   if(i < 0 || i > L->length - 1){
      L->cruz->pointer = NULL;
      L->cruz->index = -1;
   }
   L->cruz->pointer = L->front;
   L->cruz->index = 0;
   while(L->cruz->index != i){
      L->cruz->pointer = L->cruz->pointer->next;
      L->cruz->index++;
   }
}

void movePrev(List L){
   if(L == NULL){
      printf("List ERROR: calling movePrev() on a NULL List reference");
      exit(1);
   }
   if(L->cruz->pointer == NULL){
      return;
   }
   if(L->cruz->index > L->length -1){
      printf("List ERROR: calling movePrev() when L->cruz->index > L->length -1");
      exit(1);
   }
   if(L->cruz->index == 0){
      printf("movePrev() index == 0\n");
      L->cruz->pointer = NULL;
      L->cruz->index = -1;
      return;
   }
   L->cruz->pointer = L->cruz->pointer->prev;
   L->cruz->index--;
}

void moveNext(List L){
   if(L == NULL){
      printf("List ERROR: calling moveNext() on a NULL List reference\n");
      exit(1);
   }
   if(L->cruz->pointer == NULL){
      return;
   }
   if(L->cruz->index > L->length -1){
      printf("List ERROR: calling moveNext() when L->cruz->index > L->length -1\n");
      exit(1);
   }
   if(L->cruz->index == L->length - 1){
      L->cruz->pointer = NULL;
      L->cruz->index = -1;
      return;
   }
   L->cruz->pointer = L->cruz->pointer->next;
   L->cruz->index++;
}

void prepend(List L, int data){
   if(L == NULL){
      printf("List ERROR: calling prepend() on a NULL List reference");
      exit(1);
   }
   Node tmp = newNode(data);
   if(L->front == NULL){
      L->front = tmp;
      L->back = tmp;
      L->length++;
      return;
   }
   tmp->next = L->front;
   L->front->prev = tmp;
   L->front = tmp;
   if(L->cruz->index != -1){
      L->cruz->index++;
   }
   L->length++;

}

void append(List L, int data){
   if(L == NULL){
      printf("List ERROR: calling append() on a NULL List reference");
      exit(1);
   }
   Node tmp = newNode(data);
   if(L->back == NULL){
      L->front = tmp;
      L->back = tmp;
      L->length++;
      return;
   }
   tmp->prev = L->back;
   L->back->next = tmp;
   L->back = tmp;
   L->length++;
}

void insertBefore(List L, int data){
   if(L == NULL){
      printf("List ERROR: calling insertBefore() on a NULL List reference");
      exit(1);
   }
   if(L->cruz->index == -1){
      printf("List ERROR: calling insertBefore() on a NULL Cursor reference");
      exit(1);
   }
   if(L->length <= 0){
      printf("List ERROR: calling insertBefore() when length <= 0");
      exit(1);
   }
   Node tmp = newNode(data);
   //tmp->next = L->cruz->pointer;
   if(L->cruz->pointer != L->front){
      tmp->next = L->cruz->pointer;
      tmp->prev = L->cruz->pointer->prev;
      L->cruz->pointer->prev->next = tmp;
      L->cruz->pointer->prev = tmp;
   }else{
      L->cruz->pointer->prev = tmp;
      tmp->next = L->front;
      L->front = tmp;
      L->length++;
      return;
   }
   L->cruz->index++;
   L->length++;

}

void insertAfter(List L, int data){
   if(L == NULL){
      printf("List ERROR: calling insertAfter() on a NULL List reference");
      exit(1);
   }
   if(L->cruz->index == -1){
      printf("List ERROR: calling insertAfter() on a NULL Cursor reference");
      exit(1);
   }
   if(L->length <= 0){
      printf("List ERROR: calling insertAfter() when length <= 0");
      exit(1);
   }
   Node tmp = newNode(data);
   //tmp->next = L->cruz->pointer;
   if(L->cruz->pointer != L->back){
      tmp->prev = L->cruz->pointer;
      tmp->next = L->cruz->pointer->next;
      L->cruz->pointer->next->prev = tmp;
      L->cruz->pointer->next = tmp;
   }else{
      L->cruz->pointer->next = tmp;
      tmp->prev = L->back;
      L->back = tmp;
      L->length++;
      return;
   }
   L->length++;
}

void deleteFront(List L){
   if(L == NULL){
      printf("List ERROR: calling deleteFront() on a NULL List reference");
      exit(1);
   }
   if(L->length <= 0){
      printf("List ERROR: calling deleteFront() when length <= 0");
      exit(1);
   }
   if(L->cruz->index != -1){
      if(L->cruz->pointer == L->front){
         L->cruz->pointer = NULL;
         L->cruz->index = -1;
      }else{
         L->cruz->index--;
      }
   }
   if(L->front != L->back){
      L->front = L->front->next;
      freeNode(&L->front->prev);
      //L->front->prev = NULL;
   }else{
      freeNode(&L->front);
      //L->front = NULL;
      //L->back = NULL;
   }
   L->length--;
}

void deleteBack(List L){
   if(L == NULL){
      printf("List ERROR: calling deleteBack() on a NULL List reference");
      exit(1);
   }
   if(L->length <= 0){
      printf("List ERROR: calling deleteBack() when length <= 0");
      exit(1);
   }
   if(L->cruz->pointer == L->back){
      L->cruz->pointer = NULL;
      L->cruz->index = -1;
   }
   if(L->back != L->front){
      L->back = L->back->prev;
      freeNode(&L->back->next);
      L->back->prev = NULL;
   }else{
      freeNode(&L->back);
      L->back = NULL;
      L->front = NULL;
   }
   L->length--;
}

void delete(List L){
   if(L == NULL){
      printf("List ERROR: calling delete() on a NULL List reference");
      exit(1);
   }
   if(L->cruz->index == -1){
      printf("List ERROR: calling delete() on a NULL Cursor reference");
      exit(1);
   }
   if(L->length <= 0){
      printf("List ERROR: calling delete() when length <= 0");
      exit(1);
   }
   if(L->length != 0){
      if(L->cruz->pointer == L->back){
         deleteBack(L);
         L->length--;
         return;
      }
      if(L->cruz->pointer == L->front){
         deleteFront(L);
         L->length--;
         return;
      }
      Node tmp = L->cruz->pointer->prev;
      tmp->next = L->cruz->pointer->next;
      freeNode(&tmp->next->prev);
      tmp->next->prev = tmp;
   }else{
      deleteFront(L);
   }
   L->length--;
}

// Other operations -----------------------------------------------------------
void printList(FILE* out, List L){
   if(L == NULL){
      printf("List ERROR: calling printList() on a NULL List reference");
      exit(1);
   }
   if(L->front == NULL){
      printf("There are no items in the list\n");
   }
   for(Node N = L->front;N != NULL; N = N->next){
      fprintf(out, "%d ",N->data);
   }
   //fprintf(out,"\n");
}

List copyList(List L){
   if(L == NULL){
      printf("List ERROR: calling copyList() on a NULL List reference");
      exit(1);
   }
   List A = newList();
   Node N = L->front;
   while(N != NULL){
      append(A, N->data);
      N = N->next;
   }
   return A;
}



